### PR TITLE
add nfc.watch test case when nfcwatchoptions has setted

### PR DESCRIPTION
**Pull request by zhaoming0**

### CODE REVIEW

File: web-nfc/nfc/nfc_watch_nfcwatchoption_mode_webnfconly-manual.https.html [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+<div id="log"></div>
+
+<script>
+
+"use strict";
+
+setup({ explicit_timeout: true });
+
+let watchOptions = {
+  mode: "web-nfc-only"
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_nfcwatchoption_mode_webnfconly-manual.https.html]: Can you explain what is  "web-nfc-only" used for?

File: web-nfc/nfc/nfc_watch_nfcwatchoption_url-manual.https.html [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+<script>
+
+"use strict";
+
+setup({ explicit_timeout: true });
+
+let watchOptions = {
+  url: "https://w3c.github.io/*"
+}
+testWatchOptions(createTextRecord(test_text_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions.url is set.");
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_nfcwatchoption_url-manual.https.html]: Is text record type can be watched by url watchOptions?

File: web-nfc/nfc/nfc_push_and_read_JSON.manual.https.html [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ARE DISCLAIMED. IN NO EVENT SHALL INTEL CORPORATION BE LIABLE FOR ANY DIRECT,
+INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+<meta charset=utf-8>
+<title>Web NFC Test: push and read JSON </title>
```
Comment (Honry) [inline web-nfc/nfc/nfc_push_and_read_JSON.manual.https.html]: It's not a good name, as people would be confused with the test: nfc_push_nfcmessage_json-manual.https.html

**Comment by zhaoming0** [comment]
update it

File: web-nfc/nfc/nfc_watch_recordtype_json-manual.https.html [hunk 4]
```diff
@@ -49,8 +49,10 @@
 setup({ explicit_timeout: true });
 
 let watchOptions = {
-  recordType : "text"
+  recordType : "json",
+  mediaType : "application/json",
+  url : ""
 }
-testWatchOptions(createTextRecord(test_text_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions.recordType is set.");
+testWatchOptions(createJsonRecord(test_json_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions recordType 'json' is set.");
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_recordtype_json-manual.https.html]: when NFCWatchOptions recordType  is set to 'json'.

File: web-nfc/nfc/nfc_watch_recordtype_opaque-manual.https.html [hunk 5]
```diff
@@ -51,8 +49,10 @@
 setup({ explicit_timeout: true });
 
 let watchOptions = {
-  mode: "any"
+  recordType : "opaque",
+  mediaType : "application/octet-stream",
+  url : ""
 }
-testWatchOptions(createTextRecord(test_text_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions.mode is set to 'any'.");
+testWatchOptions(createOpaqueRecord(test_buffer_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions recordType 'opaque' is set.");
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_recordtype_opaque-manual.https.html]: ditto.

File: web-nfc/nfc/nfc_watch_recordtype_text-manual.https.html [hunk 6]
```diff
@@ -49,8 +49,10 @@
 setup({ explicit_timeout: true });
 
 let watchOptions = {
-  mediaType: "text/plain"
+  recordType : "text",
+  mediaType : "text/plain",
+  url : ""
 }
-testWatchOptions(createTextRecord(test_text_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions.mediaType is set.");
+testWatchOptions(createTextRecord(test_text_data), watchOptions, "Test that nfc.watch succeeds when NFCWatchOptions recordType 'text' is set.");
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_recordtype_text-manual.https.html]: Ditto.

File: web-nfc/nfc/nfc_help.js [hunk 7]
```diff
@@ -139,44 +139,45 @@ function createUrlRecord(url) {
   return createRecord("url", "text/plain", url);
 }
 
-function testNFCMessage(datas, desc) {
+function testNFCOptions(pushOptions, desc, watchOptions) {
   promise_test(t => {
-    return navigator.nfc.push({data: [{data: datas.data, recordType: datas.recordType, mediaType: datas.mediaType}]})
+    return navigator.nfc.push({data: [ { data: pushOptions.data, recordType: pushOptions.recordType, mediaType: pushOptions.mediaType}]})
```
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: It's "NFCPushMessage", you can just use “message”.

File: web-nfc/nfc/nfc_help.js [hunk 8]
```diff
@@ -139,44 +139,45 @@ function createUrlRecord(url) {
   return createRecord("url", "text/plain", url);
 }
 
-function testNFCMessage(datas, desc) {
+function testNFCOptions(pushOptions, desc, watchOptions) {
```
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: We could make watchOptions default value is null, e.g.  `function testNFCOptions(pushOptions, desc, watchOptions = null) {`

File: web-nfc/nfc/nfc_help.js [hunk 9]
```diff
@@ -139,44 +139,45 @@ function createUrlRecord(url) {
...
   promise_test(t => {
-    return navigator.nfc.push({data: [{data: datas.data, recordType: datas.recordType, mediaType: datas.mediaType}]})
+    return navigator.nfc.push({data: [ { data: pushOptions.data, recordType: pushOptions.recordType, mediaType: pushOptions.mediaType}]})
       .then(() => {
         return new Promise(resolve => {
-          navigator.nfc.watch((message) => resolve(message), {recordType: datas.recordTypes, mediaType: datas.mediaType});
+          if (watchOptions !== null && watchOptions !== undefined) {
+            navigator.nfc.watch((message) => resolve(message), watchOptions);
+          }else{
```
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: add white space before and after `else`

File: web-nfc/nfc/nfc_help.js [hunk 10]
```diff
@@ -139,44 +139,45 @@ function createUrlRecord(url) {
...
-                break;
               default:
                 assert_unreached("Invalid RecordType");
                 break;
             }
           }
-        });
-      });
-  }, desc);
+        })
```
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: Why remove the `;`?

File: web-nfc/nfc/nfc_help.js (hunk 8 above)
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: I think we don't need to rename the function name.

Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: move desc param to the last place.

File: web-nfc/nfc/nfc_watch_recordtype_json-manual.https.html [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+"use strict";
+
+setup({ explicit_timeout: true });
+
+let watchOptions = {
+  recordType : "json",
+  mediaType : "application/json",
+  url : ""
+}
+testNFCOptions(createJsonRecord(test_json_data),  "when NFCWatchOptions recordType is set to 'json'.", watchOptions);
```
Comment (Honry) [inline web-nfc/nfc/nfc_watch_recordtype_json-manual.https.html]: “Test that nfc.watch succeeds when NFCWatchOptions recordType is set to 'json'.”

File: web-nfc/nfc/nfc_help.js (hunk 7 above)
Comment (zhaoming0) [inline web-nfc/nfc/nfc_help.js]: }).then((message) => {
     for (let record of message.data) {
maybe could use "pushMessage"

File: web-nfc/nfc/nfc_help.js (hunk 9 above)
Comment (zhaoming0) [inline web-nfc/nfc/nfc_help.js]: done

File: web-nfc/nfc/nfc_help.js (hunk 7 above)
Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: sure

File: web-nfc/nfc/nfc_help.js (hunk 8 above)
Comment (zhaoming0) [inline web-nfc/nfc/nfc_help.js]: specital test point, we need add test case cover it

Comment (Honry) [inline web-nfc/nfc/nfc_help.js]: ????

**Comment by Honry** [comment]
LGTM, please cherry-pick to gh-page branch.

**Pull request merged**